crypto/evp/evp_enc.c

```diff
@@ -444,22 +456,71 @@ int EVP_CipherInit_ex2(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
                        const unsigned char *key, const unsigned char *iv,
                        int enc, const OSSL_PARAM params[])
 {
-    return evp_cipher_init_internal(ctx, cipher, NULL, key, iv, enc, params);
+    return evp_cipher_init_internal(ctx, cipher, NULL, key, iv, enc, 0, params);
 }
 
 int EVP_CipherInit(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
                    const unsigned char *key, const unsigned char *iv, int enc)
 {
     if (cipher != NULL)
         EVP_CIPHER_CTX_reset(ctx);
-    return evp_cipher_init_internal(ctx, cipher, NULL, key, iv, enc, NULL);
+    return evp_cipher_init_internal(ctx, cipher, NULL, key, iv, enc, 0, NULL);
 }
 
 int EVP_CipherInit_ex(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
                       ENGINE *impl, const unsigned char *key,
                       const unsigned char *iv, int enc)
 {
-    return evp_cipher_init_internal(ctx, cipher, impl, key, iv, enc, NULL);
+    return evp_cipher_init_internal(ctx, cipher, impl, key, iv, enc, 0, NULL);
+}
+
+int EVP_CipherPipelineInit(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
+                           const unsigned char *key, size_t numpipes,
+                           const unsigned char **iv, int enc)
+{
+    if (numpipes > EVP_MAX_PIPES) {
+        ERR_raise(ERR_LIB_EVP, EVP_R_TOO_MANY_PIPES);
+        return 0;
+    }
+
+    ctx->numpipes = numpipes;
+
+    if (!evp_cipher_init_internal(ctx, cipher, NULL, NULL, NULL, enc, 1,
+                                  NULL))
+        return 0;
+
+    /* ctx->encrypt is set by evp_cipher_init_internal */
+    if (ctx->encrypt) {
+        if (ctx->cipher->p_einit == NULL) {
+            ERR_raise(ERR_LIB_EVP, EVP_R_INITIALIZATION_ERROR);
+            return 0;
+        }
+
+        return ctx->cipher->p_einit(ctx->algctx,
+                                    key,
+                                    key == NULL ? 0
+                                                : EVP_CIPHER_CTX_get_key_length(ctx),
+                                    numpipes,
+                                    iv,
+                                    iv == NULL ? 0
+                                               : EVP_CIPHER_CTX_get_iv_length(ctx),
+                                    NULL);
+    }
+
+    if (ctx->cipher->p_dinit == NULL) {
+        ERR_raise(ERR_LIB_EVP, EVP_R_INITIALIZATION_ERROR);
+        return 0;
+    }
+
+    return ctx->cipher->p_dinit(ctx->algctx,
+                                key,
+                                key == NULL ? 0
+                                            : EVP_CIPHER_CTX_get_key_length(ctx),
+                                numpipes,
+                                iv,
+                                iv == NULL ? 0
+                                           : EVP_CIPHER_CTX_get_iv_length(ctx),
+                                NULL);
```

Comment:
Why do you copy these 2 calls from `evp_cipher_init_internal` and don't do them as a part of the modified version of evp_cipher_init_internal?

Reply:
The initialization functions for provided ciphers differ in pipeline. For encryption, `evp_cipher_init_internal` invokes `ctx->cipher->einit`, while `EVP_CipherPipelineInit` invokes `ctx->cipher->p_einit`. Also, the type of the `iv` parameter varies between the two.